docs/dev/data_validation.rst

```diff
@@ -0,0 +1,688 @@
+building on the above example would look like:
+
+.. code-block:: yaml
+
+    version: 2
+    sources:
+      - name: pudl
+        tables:
+          - name: new_table_name
+            data_tests:
+              - expect_columns_not_all_null
+              - check_row_counts_per_partition:
+                  table_name: new_table_name
+                  partition_column: report_date
+
+--------------------------------------------------------------------------------
+Defining new data validation tests
+--------------------------------------------------------------------------------
+
+* How to define a new generic test (lean on references to dbt docs when possible)
+* Focus on the things that make the PUDL use case unusual.
+* DuckDB + Parquet means we can't rely on ``adapter`` object methods (no real DB)
+* Almost all our tables are "sources" not "models"
+
+Defining Macros
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* In dbt, macros are reusable SQL snippets that can be used to simplify your tests. You
+  can define a macro once and then use it in multiple tests. This is particularly useful
+  for complex tests that require a lot of boilerplate code.
+
+Testing the Tests
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* One reason to create macros for more complex functions is that they can be
+  independently unit-tested.
+
+Creating intermediate tables for a test
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. todo::
+
+   Do we actually have any tests that do the first option below? I don't really
+   understand what it is suggesting :user:`zaneselvans`
+
+In some cases you may need to modify a table or calculate some derived values before
+you can apply a test. There are two ways to accomplish this. First, you can add the
+table as a ``source`` as described above, then create a SQL file in the ``tests/``
+directory like ``tests/{data_source}/{table_name}.yml``.  From here you can construct a
+SQL query to modify the table and execute a test on the intermediate table you've
+created. ``dbt`` expects a SQL test to be a query that returns 0 rows for a successful
+test. See the ``dbt`` `source function
+<https://docs.getdbt.com/reference/dbt-jinja-functions/source>`_ for guidance on how to
+reference a ``source`` from a SQL file.
+
+The second method is to create a `model <https://docs.getdbt.com/docs/build/models>`_
+which will produce the intermediate table you want to execute tests on. To use this
+approach, simply add a sql file to ``dbt/models/{data_source}/{table_name}/``. Now, add
+a SQL file to this directory named ``validate_{table_name}`` and define your model for
+producing the intermediate table here. Finally, add the model to the ``schema.yml`` file
+and define tests exactly as you would for a ``source`` table. See
+``models/ferc1/out_ferc1__yearly_steam_plants_fuel_by_plant_sched402`` for an example of
+this pattern.
+
+Note: when adding a model, it will be stored as a SQL ``view`` in the file
+``$PUDL_OUTPUT/pudl_dbt_tests.duckdb``.
```

Comment:
This could be a good place to note that while most other compiled tests can be run from an empty duckdb prompt, compiled tests against these intermediate models require `dubkdb $PUDL_OUTPUT/pudl_dbt_tests.duckdb`